### PR TITLE
DVGeometryMulti

### PR DESCRIPTION
**Pull request by sseraj**
## Purpose

This PR adds DVGeometryMulti, the component-based surface mesh deformation method developed by @anilyil as described in [this paper](http://websites.umich.edu/~mdolaboratory/pdf/Yildirim2021b.pdf), with some modifications by me and @yqliaohk to get it 'production ready'.

## Type of change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing

The test coverage should be pretty good. test_DVGeometryMulti.py tests the intersection functionality and derivatives. I also expanded the parameterization in test_DVConstraints.py to test all the constraints with a multi component setup.

## Checklist

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

### CODE REVIEW

**Comment by sseraj** [comment]
The failed tests are because of #89

**Comment by sseraj** [comment]
>One question: do we want to also create a PR to parallel solvers (at least adflow) that add point sets in parallel?

I'm planning on opening a PR on ADflow once this is merged. I realized that the changes to the `addPointSet` call is not backwards compatible with older pyGeo in certain cases (specifically with child FFDs), so I wanted to hold off on opening the PR until this is merged.

On a related note, I will bump the minor version of pyGeo in this PR so that any backwards incompatibility with older versions is clear in the next ADflow release.

File: pygeo/parameterization/DVGeo.py
Comment (marcomangano) [inline pygeo/parameterization/DVGeo.py:600]: So does the point set have a comm? I am not sure about how pyGeo handles parallelization.
Any plans to fix/improve this at some point?

File: pygeo/pyBlock.py
Comment (marcomangano) [inline pygeo/pyBlock.py:794]: Given that there are no additional kwargs, we could just pass `nIter` as explicit argument to `atttachPoints`

**Review by marcomangano — COMMENTED** [review]
I gave a quick look at the parts of the PR, great effort. I have two additional comments:
- It would be great to have a section of the tutorial with a working example. I get the overall idea but there are a lot of additional features (e.g. intersection management) that will require some explanation
- How is the newly added file doing in terms of code duplication? Is it going to be part of the planned code overhauling or is it already leveraging on other existing mehtods/classes?

**Comment by sseraj** [comment]
Thanks for taking the time to review this one @marcomangano; it's not an easy one. 

> It would be great to have a section of the tutorial with a working example. I get the overall idea but there are a lot of additional features (e.g. intersection management) that will require some explanation

I agree, and this is in the works. Given that this is an advanced feature and relies of pySurf, we opted to put it in the private tutorial. Have a look at the `intersection` branch for the work in progress.

> How is the newly added file doing in terms of code duplication? Is it going to be part of the planned code overhauling or is it already leveraging on other existing mehtods/classes?

DVGeometryMulti mimics the API of DVGeometry for compatibility, but there is not much code duplication. In many cases, DVGeometryMulti functions will loop over components and call the corresponding function on the DVGeometry object. See `getValues` for an example.

File: pygeo/parameterization/DVGeo.py
Comment (sseraj) [reply pygeo/parameterization/DVGeo.py:600]: I will mostly defer this question to @anilyil.

To give some context though, the motivation behind adding this is because we want to add `comm` to the `addPointSet` call in ADflow so that DVGeometryMulti can use it. I think we can remove this though because `addPointSet` already takes in arbitrary kwargs and just ignores the ones it doesn't use.

File: pygeo/pyBlock.py
Comment (sseraj) [reply pygeo/pyBlock.py:794]: This is related to the `comm` kwargs discussion above. `attachPoints` is called in `addPointSet`
https://github.com/mdolab/pygeo/blob/70ae801f137fe7c0be1d4a72214842e4c8ab0f29/pygeo/parameterization/DVGeo.py#L616-L619
We can make it can explicit argument in attachPoints, but then we simply move the unpacking up one level to `addPointSet`.

File: pygeo/parameterization/DVGeo.py
Comment (anilyil) [reply pygeo/parameterization/DVGeo.py:600]: The FFD-based DVGeos do not really need a communicator; the only parallel operation is the allreduce after the sensitibity computation, and that method takes in an optional comm. I added the comm to the addpointset call because some of the operations on the pointsets need to be comm-aware. 

There are 2 types of pointsets added to dvgeo: serial (duplicated) or distributed. The "serial" pointsets are identical across all procs and this serial computation is basically duplicated on all procs. When this is done, we dont need a comm, because all procs own the entire copy. These points are added by DVCon, so they would not get the intersection treatment in dvgeo anyways.

The parallel points are what sth like adflow adds. These are distributed among adflow's communicator and each processor gets part (or none) of the points. The FFD-based computation does not need to know the global collection of points, so we don't need a comm except until that final allreduce in the sensitivity computation. 

However, the intersection method does need to know the global collection of nodes. This is because the intersection code will need to be aware of all nodes for a particular step in the algorithm (the curve based projection and warping, sections 2.7 and 2.8 in the paper). Because of this, we are adding this optional argument; the "serial" pointsets dont use it, but the distributed pointsets do use it. On the pyGeo side, only the DVGeoMulti uses it.

Regarding absorbing it into the kwargs; if thats already there then I am fine with it since other dvgeos dont use it and we can leave it explicitly defined in the dvgeomulti code

**Comment by anilyil** [comment]
The changes besides the DVGeoMulti code itself look good to me, but someone other than me should also take a look. Thanks again for all the work here @sseraj @yqliaohk

**Comment by sseraj** [comment]
> The changes besides the DVGeoMulti code itself look good to me, but someone other than me should also take a look. 

I agree, I will leave this review to:
@marcomangano: mainly to review the parts outside of DVGeoMulti.py as you have already done, particularly the tests to see if anything can be improved there 
@yqliaohk: to review the tests and DVGeoMulti.py more closely, particularly the docstrings and code clarity

File: pygeo/pyBlock.py
Comment (marcomangano) [reply pygeo/pyBlock.py:794]: I see, this is fine as is then

File: pygeo/parameterization/DVGeo.py
Comment (marcomangano) [reply pygeo/parameterization/DVGeo.py:598]: Thanks for the explanation! I see where this goes, but I am still not sure if we want to add a "dead" attribute to this method. I am also wondering if having functions with the same name and "slightly" different behavior in different files (e.g., it looks like we now have three `addPointSet` in pygeo) can cause problems. I assume that once you define the DVGeo object then there are no ambiguities, I am just not sure if this is good practice

Comment (marcomangano) [reply pygeo/parameterization/DVGeo.py:598]: I also noticed that for consistency you should add `compNames` here. Do we want to add a warning in case these two arguments are passed in here by mistake. something like "method called with XXX attribute, but XXX is not implemented on DVGeo so it will be ignored"

File: tests/reg_tests/test_DVGeometryMulti.py
Comment (marcomangano) [inline tests/reg_tests/test_DVGeometryMulti.py:176]: It took me a bit to figure out what is going on here, we are basically getting "individual" derivative outputs for every node coordinate, right? I am not sure if it is worth adding some clarification, especially on a test script though

**Review by marcomangano — DISMISSED** [review]
Just some additional questions on the test for now, but overall is looking good and I do not have major recommendations

File: pygeo/parameterization/DVGeo.py
Comment (sseraj) [reply pygeo/parameterization/DVGeo.py:598]: I removed `comm` from this function and similarly will not add `compNames`.

For the warning, I think it is rare for the user to call `addPointSet` directly. This is mainly used by DVCon and our solvers. DVCon now passes `compNames` for every point set based constraint, so having a warning every time would be annoying.

As for whether this is good practice, I can't really say. The reason we do it is mainly to reduce code duplication in DVCon. It would be cumbersome to add code to each constraint to check what type of DVGeo was passed in and then use the appropriate `addPointSet` call.

File: tests/reg_tests/test_DVGeometryMulti.py
Comment (sseraj) [reply tests/reg_tests/test_DVGeometryMulti.py:156]: No reason really. The twists are only applied on one end of each box, so the intersection curve is changing. It's not a rigid body rotation if that was the concern.

File: tests/reg_tests/test_DVGeometryMulti.py
Comment (sseraj) [reply tests/reg_tests/test_DVGeometryMulti.py:176]: Yes, that's correct. I will add a comment to clarify this.

File: tests/reg_tests/test_DVGeometryMulti.py
Comment (marcomangano) [reply tests/reg_tests/test_DVGeometryMulti.py:47]: Ok I see, so this will be covered in the documentation at some point :)

**Review by yqliaohk — DISMISSED** [review]
Looks good to me. I also ran the tests locally and they all passed. Thanks for the work!

File: pygeo/parameterization/DVGeoMulti.py
Comment (sseraj) [reply pygeo/parameterization/DVGeoMulti.py]: Good point. I updated the docstring.

File: tests/reg_tests/test_DVConstraints.py
Comment (sseraj) [reply tests/reg_tests/test_DVConstraints.py]: That's correct. Let me know if the updated description is clearer.

Comment (yqliaohk) [reply tests/reg_tests/test_DVConstraints.py]: I see, the updated description is clearer, but why is a not-intersected case chosen as the test case here?

Comment (sseraj) [reply tests/reg_tests/test_DVConstraints.py]: The idea behind parameterizing the DVConstraints tests is to use the same reference values for all the tests. This checks that the function values and gradients for a single component of DVGeometryMulti are the same as one instance of DVGeometry.

Also, `applyIC` is False for all constraint point sets. This is because we want constraints like thickness and volume to follow the component geometry, not the surface mesh. So even if the components intersected, there would be no intersection treatment applied to the point sets.

Comment (yqliaohk) [reply tests/reg_tests/test_DVConstraints.py]: I see. Thanks for the clarification!

**Review by yqliaohk — APPROVED** [review]

**Review by marcomangano — APPROVED** [review]

**Pull request merged**
